### ISSUE TITLE
[Issue] Authentication problem when trying to delete tag

### ISSUE THREAD

**Issue by axeluhl** [regclient/regclient#143]
### Current Behavior

I have a private registry based on the ``registry`` image that runs behind an Apache / httpd reverse proxy where the reverse proxy authenticates users with basic authentication and is itself the target of an AWS ALB which does SSL termination. Listing tags with regctl works fine based on the credentials in .docker/config.json:

```
    $ ./regctl tag ls docker.example.com:443/myrepo
    build-202201131607
    latest
```

Trying to delete fails:

```
    $ ./regctl tag delete docker.example.com:443/myrepo:build-202201131607
    Failed sending dummy config to delete docker.example.com:443/myrepo:build-202201131607: Failed to send blob (put), digest sha256:f297e5c987405715cc81c6a224da00f930ba8f615e0723d7ad63f93d55b9b3ab, ref docker.example.com:443/myrepo:build-202201131607: Unable to reuse reader
```

Looking at debug output with ``-v debug`` shows a line like the following:

```
DEBU[0000] Delete tag                                    host="docker.example.com:443" repository=myrepo tag=build-202201131607
DEBU[0000] Sending request                               method=DELETE url="https://docker.example.com:443/v2/repo/manifests/build-202201131607" withAuth=false
DEBU[0000] Auth request parsed                           challenge="[{basic map[realm:Docker Registry Authentication]}]"
DEBU[0000] Sending request                               method=DELETE url="https://docker.example.com:443/v2/myrepo/manifests/build-202201131607" withAuth=true
DEBU[0000] Unexpected status                             Body="{\"errors\":[{\"code\":\"DIGEST_INVALID\",\"message\":\"provided digest did not match uploaded content\"}]}\n" Status="400 Bad Request" URL="https://docker.example.com:443/v2/myrepo/manifests/build-202201131607"
```

### Expected Behavior

I would have expected the authentication to be applied at the first DELETE request already, not provoking the auth request and hence most likely deleting the tag properly.

### Steps To Reproduce

Run a registry not doing local auth behind an Apache reverse proxy doing basic auth based on .htpasswd. Log on to the registry, push an image with a tag, then try to delete the tag.

### Version

```{
  "VCSRef": "6a1a13c410f734f5e18a6032936bc6764814eae7",
  "VCSTag": "v0.3.10"
}
```

### Environment

Running the local binary on Ubuntu Linux 16.04LTS. The registry is based on the ``registry`` image, sha256:b8604a3fe8543c9e6afc29550de05b36cd162a97aa9b2833864ea8a5be11f3e2, currently the latest registry:2 image.

In case you'd like to test/reproduce against the actual registry (not the fake ``docker.example.com`` URL) let me know and we can discuss creating credentials for you for that registry.

**Comment by sudo-bmitch** [comment]
Hi @axeluhl

> I would have expected the authentication to be applied at the first DELETE request already, not provoking the auth request and hence most likely deleting the tag properly.

This is just how authentication works, we don't know how the server expects authentication so we wait for the request with details on the realm, whether bearer tokens are needed, and other details, to generate the auth request. Docker does this by pinging just the /v2/ API, and I'm going straight to the target API. The latter method saves a hop for sites that don't have auth requirements on specific API's (e.g. pull requests).

> DEBU[0000] Unexpected status                             Body="{\"errors\":[{\"code\":\"DIGEST_INVALID\",\"message\":\"provided digest did not match uploaded content\"}]}\n" Status="400 Bad Request" URL="https://docker.example.com:443/v2/myrepo/manifests/build-202201131607"

This one is expected, the 400 bad request is seen from any registry that doesn't support the DELETE on a tag, which is just about all of them. OCI added the API in distribution-spec v1, and it's going to take time for everyone to support it.

>     Failed sending dummy config to delete docker.example.com:443/myrepo:build-202201131607: Failed to send blob (put), digest sha256:f297e5c987405715cc81c6a224da00f930ba8f615e0723d7ad63f93d55b9b3ab, ref docker.example.com:443/myrepo:build-202201131607: Unable to reuse reader

This one is going to be the issue, and probably the debug line or two before it if you have them. We send a fake image to the registry to implement the tag delete on registries without the new API. This way we can delete that fake manifest knowing no other tags point to it (normally deleting a manifest deletes all tags pointing to the same manifest, which is why `tag rm` and `image rm` are different).

Send over those other debug logs so we can better track this down. And if you have time, feel free to give the latest CI build a try in case something got fixed since the last release. Those can be found linked on the GHA page for the Go builds: https://github.com/regclient/regclient/actions/runs/1710600264

**Comment by axeluhl** [comment]
Here you go:

```
DEBU[0000] Loading docker cred                           host="docker.example.com:443" name="docker.example.com:443" pass-set=true token-set=false user=***
DEBU[0000] Loading docker cred                           host=registry-1.docker.io name=docker.io pass-set=true token-set=false user=***
DEBU[0000] Loading docker cred                           host=docker.example.com name=docker.example.com pass-set=true token-set=false user=***
DEBU[0000] regclient initialized                        
DEBU[0000] Delete tag                                    host="docker.example.com:443" repository=myrepo tag=build-202201131400
DEBU[0000] Sending request                               method=DELETE url="https://docker.example.com:443/v2/myrepo/manifests/build-202201131400" withAuth=false
DEBU[0000] Auth request parsed                           challenge="[{basic map[realm:Docker Registry Authentication]}]"
DEBU[0000] Sending request                               method=DELETE url="https://docker.example.com:443/v2/myrepo/manifests/build-202201131400" withAuth=true
DEBU[0000] Unexpected status                             Body="{\"errors\":[{\"code\":\"DIGEST_INVALID\",\"message\":\"provided digest did not match uploaded content\"}]}\n" Status="400 Bad Request" URL="https://docker.example.com:443/v2/myrepo/manifests/build-202201131400"
DEBU[0000] EOF before reading all content, retrying      contentLen=98 curRead=0
DEBU[0000] Sending request                               method=HEAD url="https://docker.example.com:443/v2/myrepo/manifests/build-202201131400" withAuth=true
DEBU[0000] Sending dummy manifest to replace tag         ref="docker.example.com:443/myrepo:build-202201131400"
DEBU[0000] Sending request                               method=POST url="https://docker.example.com:443/v2/myrepo/blobs/uploads/?mount=sha256%3A0c3cfae7f0c86270e4c8218022d8d444b3024c443152c5b4a3946361103baf6c" withAuth=true
DEBU[0000] Sending request                               method=PUT url="https://docker.example.com/v2/myrepo/blobs/uploads/eb166be0-7bf7-41fa-8c4f-12701a6d8006?_state=XHSO5re5K-Zi8K_KXSgCWoO4jTPi-Y0j2y9HcopUblJ7Ik5hbWUiOiJzYXBzYWlsaW5nIiwiVVVJRCI6ImViMTY2YmUwLTdiZjctNDFmYS04YzRmLTEyNzAxYTZkODAwNiIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyMi0wMS0xOVQxNTowODo1Ni40MzA4NDg3NThaIn0%3D&digest=sha256%3A0c3cfae7f0c86270e4c8218022d8d444b3024c443152c5b4a3946361103baf6c" withAuth=false
DEBU[0000] Auth request parsed                           challenge="[{basic map[realm:Docker Registry Authentication]}]"
Failed sending dummy config to delete docker.example.com:443/myrepo:build-202201131400: Failed to send blob (put), digest sha256:0c3cfae7f0c86270e4c8218022d8d444b3024c443152c5b4a3946361103baf6c, ref docker.example.com:443/myrepo:build-202201131400: Unable to reuse reader
```

In the meantime I can give the latest CI build a try.

**Comment by axeluhl** [comment]
With version
{
  "VCSRef": "92a844f8817ffd5ca5a3511b52d2804c35d92269",
  "VCSTag": ""
}
I'm getting this:
```
DEBU[0000] Loading docker cred                           host=docker.example.com name=docker.example.com pass-set=true tls=enabled token-set=false user=***
DEBU[0000] Loading docker cred                           host="docker.example.com:443" name="docker.example.com:443" pass-set=true tls=enabled token-set=false user=***
DEBU[0000] Loading docker cred                           host=registry-1.docker.io name=docker.io pass-set=true tls=enabled token-set=false user=***
DEBU[0000] regclient initialized
DEBU[0000] Delete tag                                    host="docker.example.com:443" repository=myrepo tag=build-202201131400
DEBU[0000] Auth request parsed                           challenge="[{basic map[realm:Docker Registry Authentication]}]"
DEBU[0000] Request failed                                Status="Bad Request" URL="https://docker.example.com:443/v2/myrepo/manifests/build-202201131400"
DEBU[0000] Sending dummy manifest to replace tag         ref="docker.example.com:443/myrepo:build-202201131400"
DEBU[0000] Auth request parsed                           challenge="[{basic map[realm:Docker Registry Authentication]}]"
Failed sending dummy config to delete docker.example.com:443/myrepo:build-202201131400: Failed to send blob (put), digest sha256:38ffbf3cb1027b023ddd49e25d1b5b69ac505a03f78de0c230485531d9fea731, ref docker.example.com:443/myrepo:build-202201131400: Unable to reuse reader
```

**Comment by sudo-bmitch** [comment]
Interesting. I think the actual error is getting masked by the retry logic, so I've got some work to better expose the error message in these cases. It feels like the repository may be read only to this user. Can you try pushing other content to the registry? Even copy a tag to a new name may help, but preferably something that adds a new layer or image config would trigger the same blob API. I'll also dig more into the possible causes and how to improve the errors.

**Comment by axeluhl** [comment]
The repo is certainly not read-only to the user. Using regular ``docker push`` commands works just fine.

**Comment by sudo-bmitch** [comment]
One other thought, you have credentials both with and without the 443 port. What happens if you remove the ":443" port from the "tag rm" command? Docker may be doing this internally since that's the default port with https.

**Comment by axeluhl** [comment]
Authentication doesn't work without the :443 which is because in the .docker/config.json file the :443 is part of the URL that serves as the key for the credentials.

**Comment by sudo-bmitch** [comment]
In that case, try doing a `docker logout` on the host without the port so there's only one entry in the config.json file. Then double check that the push still works.

**Comment by sudo-bmitch** [comment]
I think I'm seeing what happened. The blob upload is going to the hostname without the 443. This happens with how the blob API gives a new location to upload, and the 443 is getting stripped from that URL since it's the default for https (it's the same when you talk to public registries like docker hub, ghcr, etc, they're all on 443). And this code doesn't send creds immediately since that gets treated as a separate host. The bigger issue on my side is the temporary blob can't be reused for the resend, and that should be an easy code fix.

**Comment by sudo-bmitch** [comment]
@axeluhl The fix was only 1 line, but adding tests for this scenario took a bit to create your situation where the registry was on docker.example.com:443, the blob upload host was docker.example.com (no port), and both with auth. Have a look at the build artifacts here to give it a try: https://github.com/regclient/regclient/actions/runs/1731483981

**Comment by sudo-bmitch** [comment]
This update has been included in the 0.4.0 release.
